Move common js_api file to main js_api directory and add tests

findolor opened the pull request:
<!-- Thanks for your Pull Request, please read the contributing guidelines before submitting. -->

## Motivation

<!--
Explain the context and why you're making that change. What is the problem
you're trying to solve? In some cases there is not a problem and this can be
thought of as being the motivation for your change.
-->

See issue: #1621

## Solution

<!--
Summarize the solution and provide any necessary context needed to understand
the code change.
-->

## Checks
<!-- It's important you've done these, or your PR will not be considered for review -->
By submitting this for review, I'm confirming I've done the following:
- [x] made this PR as small as possible
- [x] unit-tested any new functionality
- [x] linked any relevant issues or PRs
- ~[ ] included screenshots (if this involves a front-end change)~

fix #1621

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Added a new WebAssembly (WASM) module for generating calldata for adding and removing orders, now accessible via JavaScript bindings.
  - Exposed new functions for generating add/remove order calldata from JavaScript, with improved error handling and user-friendly error messages.

- **Bug Fixes**
  - Improved error handling in the order removal modal to display clear messages if calldata generation fails.

- **Refactor**
  - Unified and simplified WASM trait derivations and macro usage across subgraph types for better interoperability.
  - Reorganized test helpers and conditional module imports for improved maintainability.

- **Tests**
  - Added comprehensive tests for the new WASM bindings, covering both successful and error scenarios.
  - Updated and migrated tests to reflect the new structure and error handling mechanisms.
  - Simplified and refocused existing test suites to align with the updated API.

- **Chores**
  - Updated test data and configuration samples to match new expected formats and values.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

findolor commented on packages/orderbook/test/js_api/common.test.ts:
```diff
@@ -0,0 +1,170 @@
+:;
+`;
+
+	it('should get correct calldata', async () => {
+		// mock calls
+		// iInterpreter() call
+		await mockServer
+			.forPost('/rpc-url')
+			.withBodyIncluding('0xf0cfdd37')
+			.thenSendJsonRpcResult(`0x${'0'.repeat(24) + '1'.repeat(40)}`);
+		// iStore() call
+		await mockServer
+			.forPost('/rpc-url')
+			.withBodyIncluding('0xc19423bc')
+			.thenSendJsonRpcResult(`0x${'0'.repeat(24) + '2'.repeat(40)}`);
+		// iParser() call
+		await mockServer
+			.forPost('/rpc-url')
+			.withBodyIncluding('0x24376855')
+			.thenSendJsonRpcResult(`0x${'0'.repeat(24) + '3'.repeat(40)}`);
+		// parse2() call
+		await mockServer
+			.forPost('/rpc-url')
+			.withBodyIncluding('0xa3869e14')
+			// 0x1234 encoded bytes
+			.thenSendJsonRpcResult(
+				'0x000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000021234000000000000000000000000000000000000000000000000000000000000'
+			);
+
+		const result = await getAddOrderCalldata(dotrain, 'some-deployment');
+		if (result.error) assert.fail('expected no error');
+		assert.equal(result.value.length, 2314);
+	});
```
No need to do this

findolor commented on crates/js_api/src/common/mod.rs:
```diff
@@ -0,0 +1,360 @@
+mod tests {
+    use super::*;
+
+    #[cfg(target_family = "wasm")]
+    mod wasm_tests {
+        use super::*;
+        use alloy::{
+            primitives::{Address, FixedBytes, U256},
+            sol_types::SolCall,
+        };
+        use rain_orderbook_bindings::IOrderBookV4::{removeOrder2Call, EvaluableV3, OrderV3, IO};
```
No need to do this

findolor commented on crates/js_api/src/common/mod.rs:
```diff
@@ -0,0 +1,360 @@
+use thiserror::Error;
+use wasm_bindgen_utils::{impl_wasm_traits, prelude::*, wasm_export};
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Tsify)]
+pub struct AddOrderCalldata(#[tsify(type = "string")] Bytes);
+impl_wasm_traits!(AddOrderCalldata);
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Tsify)]
+pub struct RemoveOrderCalldata(#[tsify(type = "string")] Bytes);
+impl_wasm_traits!(RemoveOrderCalldata);
+
```
No need to do this for now

findolor commented on crates/js_api/src/common/mod.rs:
```diff
@@ -0,0 +1,360 @@
+    }
+}
+
+/// Get addOrder() calldata from a given dotrain text and deployment key from its frontmatter
+#[wasm_export(
+    js_name = "getAddOrderCalldata",
+    unchecked_return_type = "AddOrderCalldata"
+)]
+pub async fn get_add_order_calldata(
+    dotrain: &str,
+    deployment: &str,
+) -> Result<AddOrderCalldata, Error> {
+    let config: Config = parse_frontmatter(dotrain.to_string()).await?.try_into()?;
+    let deployment_ref = config
+        .deployments
+        .get(deployment)
+        .ok_or(Error::UndefinedDeployment)?;
+    let add_order_args =
+        AddOrderArgs::new_from_deployment(dotrain.to_string(), deployment_ref.deref().clone())
+            .await?;
+
+    let tx_args = TransactionArgs {
+        rpc_url: deployment_ref.scenario.deployer.network.rpc.to_string(),
+        ..Default::default()
+    };
+    let calldata = add_order_args.get_add_order_calldata(tx_args).await?;
+    Ok(AddOrderCalldata(Bytes::copy_from_slice(&calldata)))
+}
```
This is okay in our case

findolor commented on crates/js_api/src/common/mod.rs:
```diff
@@ -0,0 +1,360 @@
+    js_name = "getRemoveOrderCalldata",
+    unchecked_return_type = "RemoveOrderCalldata"
+)]
+pub async fn get_remove_order_calldata(order: SgOrder) -> Result<RemoveOrderCalldata, Error> {
+    let remove_order_args = RemoveOrderArgs { order };
+    let calldata = remove_order_args.get_rm_order_calldata().await?;
+    Ok(RemoveOrderCalldata(Bytes::copy_from_slice(&calldata)))
+}
```
No need to do this

findolor commented on crates/js_api/src/common/mod.rs:
```diff
@@ -0,0 +1,361 @@
+};
+use rain_orderbook_subgraph_client::types::common::SgOrder;
+use serde::{Deserialize, Serialize};
+use std::ops::Deref;
+use thiserror::Error;
```
No need to do this

findolor commented on crates/js_api/src/common/mod.rs:
```diff
@@ -0,0 +1,361 @@
+    js_name = "getAddOrderCalldata",
+    unchecked_return_type = "AddOrderCalldata"
+)]
+pub async fn get_add_order_calldata(
+    dotrain: &str,
+    deployment: &str,
+) -> Result<AddOrderCalldata, Error> {
+    let config: Config = parse_frontmatter(dotrain.to_string()).await?.try_into()?;
+    let deployment_ref = config
+        .deployments
+        .get(deployment)
+        .ok_or(Error::UndefinedDeployment)?;
+    let add_order_args =
+        AddOrderArgs::new_from_deployment(dotrain.to_string(), deployment_ref.deref().clone())
+            .await?;
```
No need to do this

hardyjosh commented on Cargo.toml:
```diff
@@ -53,7 +53,7 @@ test-context = "0.3.0"
 rain-erc = { git = "https://github.com/rainlanguage/rain.erc", rev = "0106e645ebd49334addc698c5aad9a85370eb54d" }
 rain-error-decoding = { git = "https://github.com/rainlanguage/rain.error", rev = "72d9577fdaf7135113847027ba951f9a43b41827" }
-wasm-bindgen-utils = "0.0.7"
+wasm-bindgen-utils = { git = "https://github.com/rainlanguage/rain.wasm", rev = "4379d24a671742a58ec5cdab2c39fecbda59619d" }
```
@findolor is there a reason we're moving away from a published version to a git dependency?

findolor replied:
@hardyjosh that's because the wasm export was not published at the time of this PRs creation

hardyjosh commented:
@findolor i'm a bit confused by this PR - the issue is just about missing tests, but then there are a bunch of other changes

findolor commented:
@hardyjosh while writing the tests, i've also implemented the wasm macro and moved the logic to be inside the js api. that's why there are other changes. otherwise we needed to do a second round of PR to update the tests again with wasm export in mind

0xgleb commented on packages/orderbook/test/js_api/common.test.ts:
```diff
@@ -0,0 +1,170 @@
+	});
+
+	it('should get remove order calldata', async () => {
+		let order: SgOrder = {
+			id: '1',
+			orderBytes:
+				'0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000006171c21b2e553c59a64d1337211b77c367cefe5d00000000000000000000000000000000000000000000000000000000000000a000000000000000000000000000000000000000000000000000000000000001c00000000000000000000000000000000000000000000000000000000000000240000000000000000000000000000000000000000000000000000000000000000100000000 ...
+			orderHash: '',
+			owner: '',
+			active: true,
+			inputs: [],
+			outputs: [],
+			meta: undefined,
+			orderbook: {
+				id: '1'
+			},
+			addEvents: [],
+			timestampAdded: '0',
+			trades: [],
+			removeEvents: []
+		};
```
Seems fine for a test

0xgleb commented on crates/js_api/src/common/mod.rs:
We should probably have non-happy path tests

0xgleb reviewed (changes requested):
Would be good to add non-happy path test cases in Rust. Aside from that everything looks good

findolor commented on crates/js_api/src/common/mod.rs:
```diff
@@ -0,0 +1,412 @@
+    }
+}
+
+impl From<Error> for JsValue {
+    fn from(value: Error) -> Self {
+        JsError::new(&value.to_string()).into()
+    }
+}
```
no need to do this as we are returning a custom error variant that has both the original and readable message

findolor commented on crates/js_api/src/common/mod.rs:
```diff
@@ -0,0 +1,412 @@
+use thiserror::Error;
+use wasm_bindgen_utils::{impl_wasm_traits, prelude::*, wasm_export};
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Tsify)]
+pub struct AddOrderCalldata(#[tsify(type = "string")] Bytes);
+impl_wasm_traits!(AddOrderCalldata);
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Tsify)]
+pub struct RemoveOrderCalldata(#[tsify(type = "string")] Bytes);
+impl_wasm_traits!(RemoveOrderCalldata);
+
```
no need to do this. things work the way we expect at the moment

0xgleb reviewed (approved):
✅

findolor commented on packages/webapp/src/lib/components/OrderRemoveModal.svelte:
```diff
@@ -24,11 +24,14 @@
 
 	async function handleTransaction() {
-		const removeOrderCalldata = await getRemoveOrderCalldata(args.order);
+		const res = await getRemoveOrderCalldata(args.order);
+		if (res.error) {
+			throw new Error(res.error.msg);
+		}
 		transactionStore.handleRemoveOrderTransaction({
 			config: $wagmiConfig,
 			...args,
-			removeOrderCalldata
+			removeOrderCalldata: res.value
 		});
```
this error is handled by another functions so no need to use what you are suggesting

findolor commented on packages/webapp/src/__tests__/OrderRemoveModal.test.ts:
```diff
@@ -31,6 +32,9 @@ describe('OrderRemoveModal', () => {
 	beforeEach(() => {
 		vi.clearAllMocks();
 		vi.resetAllMocks();
+		(getRemoveOrderCalldata as Mock).mockResolvedValue({
+			value: '0x123'
+		});
 	});
```
no need to do this
